### DISCUSSION TITLE
Would it make sense to scope bibtex formatting settings by resource?

### DISCUSSION THREAD

**Discussion by bersbersbers** [James-Yu/LaTeX-Workshop#3191] [Q&A] [answered]
Related to #3082, I'd like to convert BibTeX formattings settings scopes to `resource`. This would involve (I guess) generating `BibtexFormatConfig`s on the fly, here:
https://github.com/James-Yu/LaTeX-Workshop/blob/49f049177a44a87b02941e6fb66209e0954a0831/src/providers/bibtexcompletion.ts#L58
https://github.com/James-Yu/LaTeX-Workshop/blob/49f049177a44a87b02941e6fb66209e0954a0831/src/providers/bibtexcompletion.ts#L60
as well as in several places in
https://github.com/James-Yu/LaTeX-Workshop/blob/d608023b862e0bc1d39735cc753061a7ffd25036/src/providers/bibtexformatterlib/bibtexutils.ts

The only downside I can think of is performance which (if relevant) might be remedied with caching.

I wonder if the maintainers would be open to a PR.

**Comment by jlelong** [comment]
The difficulty is not so much to generate `BibtexFormatConfig` on the fly but rather to take appropriate actions when the current workspace (ie. the workspace of the active text editor) changes.

You point out the lines

https://github.com/James-Yu/LaTeX-Workshop/blob/fa516ba977293305e3a9b83446d7905633974e20/src/providers/bibtexcompletion.ts#L57-L61

Note they are called through https://github.com/James-Yu/LaTeX-Workshop/blob/fa516ba977293305e3a9b83446d7905633974e20/src/providers/bibtexcompletion.ts#L47 in the constructor of `BibtexCompleter` to build the default suggestion lists once and for all. We would need a way to rebuild these lists whenever the current workspace changes. As there is no notion of active workspace in VS Code (see https://code.visualstudio.com/api/references/vscode-api#WorkspaceFolder), we would need to monitor it inside the extension along with the root file. It may be possible but this is a huge effort and I am not sure yet about the extra maintenance cost.

**Reply by bersbersbers** [reply]
Thanks for the explanation - a couple of additional thoughts.

So your explanation implies one would need not only generate the `BibtexFormatConfig` on the fly (e.g., based on the `document` argument of `provideCompletionItems`), but also need to move the `config`-dependent formatting in `entryToCompletion` to `provideCompletionItems` - in particular, these lines:
https://github.com/James-Yu/LaTeX-Workshop/blob/fa516ba977293305e3a9b83446d7905633974e20/src/providers/bibtexcompletion.ts#L94-L102

and thus format completion items on the fly based on a config for the current resource (which is `document`). Still sounds doable, but I see how this has the potential of generating a less smooth experience compared to preformatting completion items "once and for all".

I am not sure the notion of "monitoring an active workspace" is required if the scope of is `resource`. What one might use, by contrast, is `onDidChangeActiveTextEditor` (https://code.visualstudio.com/api/references/vscode-api#window) which could be used to pre-format the completion items (as it is now, but then using the formatting for this document's BibTeX formatting config) so that the actual completion remains as instantaneous as it is.

By the way, my current understanding is that changes to the BibTeX formatting config settings are not picked up by the `BibtexCompleter` until the extension is reloaded, which may be considered an issue that could be solved along the lines I pointed out (but I may be missing something).

**Reply by jlelong** [reply]
> Still sounds doable, but I see how this has the potential of generating a less smooth experience compared to preformatting completion items "once and for all".

I am against recomputing  the complete suggestion list every time completion is triggered because of the computational overhead. A "greener" approach is needed.

> By the way, my current understanding is that changes to the BibTeX formatting config settings are not picked up by the `BibtexCompleter` until the extension is reloaded, which may be considered an issue that could be solved along the lines I pointed out (but I may be missing something).

I added to the description of these settings that any change requires to reload VS Code, see 1ebddded. I agree that it would be a great improvement but it would require a huge implementation effort. I have to think deeper about it.

> I am not sure the notion of "monitoring an active workspace" is required if the scope of is `resource`. What one might use, by contrast, is `onDidChangeActiveTextEditor` (https://code.visualstudio.com/api/references/vscode-api#window) which could be used to pre-format the completion items (as it is now, but then using the formatting for this document's BibTeX formatting config) so that the actual completion remains as instantaneous as it is.

Of course, we could only monitor the change of active text editor but it would induce a lot of unnecessary computations of the default completion items. The idea behind "monitoring an active workspace" is to avoid these unused computations.

**Reply by bersbersbers** [reply]
> The idea behind "monitoring an active workspace" is to avoid these unused computations.

Understood. The question is if monitoring the active workspace is sufficient if the the settings did in fact become `resource`-scoped, in which case they *might* be expected to differ from file to file (note that this is not the use case I had in mind with #3082, but other people might).

Anyway, I see the difficulties now, and I'll continue to think about this as well - thanks a lot for the discussion so far!

**Reply by jlelong** [reply]
> The question is if monitoring the active workspace is sufficient if the the settings did in fact become `resource`-scoped, in which case they _might_ be expected to differ from file to file

When using a `resource` scope, the settings associated to a file are actually those tight to the workspace containing the file. So yes, it is sufficient to monitor an active workspace change. This is indeed quite complicated to implement and it is orthogonal to the logic currently used by VS Code for multi-root workspace, see https://github.com/yzhang-gh/vscode-markdown/issues/1092#issuecomment-1068073980.

Note that this would still not enable us to make any bibtex settings change effective without reloading the window.

**Comment by jlelong** [comment, marked as answer]
PR #3219 closes this discussion. Now, all `bibtex` related settings have the `resource` scope and any configuration change is immediately effective without reloading.

**Reply by bersbersbers** [reply]
Excellent, this solves several issues at once. Thanks a lot!

**Reply by jlelong** [reply]
This is shipped with version `8.24.1`, which has just been released.
